zvariant/src/file_path.rs

```diff
@@ -0,0 +1,106 @@
+use std::{ffi::{OsStr, OsString}, path::{Path, PathBuf}};
+
+use serde::{de::{self, Visitor}, Deserialize, Deserializer, Serialize, Serializer};
+use crate::de::Deserializer as ZbusDeserializer;
+
+use crate::Type;
+
+/// A file name represented as a nul-terminated byte array.
+#[derive(Type, Debug, Default, PartialEq)]
+#[zvariant(signature = "ay")]
+pub struct FilePath(OsString);
+
+impl From<&Path> for FilePath {
+    fn from(value: &Path) -> Self {
+        Self(value.as_os_str().to_os_string())
+    }
+}
+
+impl From<PathBuf> for FilePath {
+    fn from(value: PathBuf) -> Self {
+        Self(value.as_os_str().to_os_string())
+    }
+}
+
+impl<'de> Deserialize<'de> for FilePath {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: Deserializer<'de>
+    {
+        struct FilePathVisitor;
+        impl<'de> Visitor<'de> for FilePathVisitor {
+            type Value = FilePath;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                formatter.write_str("unable to deserialize FilePath")
```

Comment:
The erroring out itself already conveys the failure to deserialize FilePath. This string is about telling what we were expecting: "a byte array".